File: js/src/encode.ts
import { ByteWriter } from "./buffer";
import { MAGIC, VERSION, Tag, KeyKind, ElementType } from "./format";
import type { TypeExtension } from "./extension";

// Explicit contents for weak collections, since WeakMap/WeakSet are not
// enumerable per spec. The caller supplies the entries they are holding.
export interface WeakProvider {
  weakMapEntries?: (wm: WeakMap<object, unknown>) => Array<[object, unknown]>;
  weakSetValues?: (ws: WeakSet<object>) => object[];
}

export interface EncodeOptions {
  provider?: WeakProvider;
  /** Extension types for values the core would otherwise reject. */
  types?: TypeExtension[];
}

interface Node {
  write(w: ByteWriter, idOf: (v: unknown) => number): void;
}

const wellKnownSymbols = new Map<symbol, string>(
  (Object.getOwnPropertyNames(Symbol) as Array<keyof typeof Symbol>)
    .filter((n) => typeof (Symbol as any)[n] === "symbol")
    .map((n) => [(Symbol as any)[n] as symbol, String(n)]),
);

export function encode(root: unknown, options: EncodeOptions = {}): Uint8Array {
  const provider: WeakProvider = options.provider ?? {};
  const types = options.types ?? [];
  const heap: Node[] = [];
  // identity map for objects/symbols (reference types)
  const ids = new Map<unknown, number>();

  function intern(v: unknown): number {
    const existing = ids.get(v);
    if (existing !== undefined) return existing;
    const idx = heap.length;
    ids.set(v, idx);
    heap.push(null as unknown as Node); // placeholder to reserve index (cycles)
    heap[idx] = build(v);
    return idx;
  }

  function leaf(write: Node["write"]): Node {
    return { write };
  }

  function build(v: unknown): Node {
    switch (typeof v) {
      case "undefined":
        return leaf((w) => w.u8(Tag.Undefined));
      case "boolean":
        return leaf((w) => w.u8(v ? Tag.BoolTrue : Tag.BoolFalse));
      case "bigint":
        return leaf((w) => {
          w.u8(Tag.BigInt);
          w.u8(v < 0n ? 1 : 0);
          w.uvarint(v < 0n ? -v : v);
        });
      case "string":
        return leaf((w) => {
          w.u8(Tag.String);
          w.str(v);
        });
      case "number":
        return buildNumber(v);
      case "symbol":
        return buildSymbol(v);
      case "object":
        if (v === null) return leaf((w) => w.u8(Tag.Null));
        return buildObject(v as object);
      case "function":
        throw new Error("functions are out of scope");
      default:
        throw new Error("unsupported type: " + typeof v);
    }
  }

  function buildNumber(v: number): Node {
    // Use Int when it is an exact integer (and not -0), else Float to keep
    // NaN / -0 / ±Infinity / fractional values bit-faithful.
    const isExactInt = Number.isInteger(v) && !Object.is(v, -0);
    if (isExactInt) {
      return leaf((w) => {
        w.u8(Tag.Int);
        w.svarint(BigInt(v));
      });
    }
    return leaf((w) => {
      w.u8(Tag.Float);
      w.f64(v);
    });
  }

  function buildSymbol(sym: symbol): Node {
    const wk = wellKnownSymbols.get(sym);
    if (wk)
      return leaf((w) => {
        w.u8(Tag.SymbolWellKnown);
        w.str(wk);
      });
    const key = Symbol.keyFor(sym);
    if (key !== undefined) {
      return leaf((w) => {
        w.u8(Tag.SymbolRegistered);
        w.str(key);
      });
    }
    const desc = sym.description ?? "";
    return leaf((w) => {
      w.u8(Tag.SymbolUnique);
      w.str(desc);
    });
  }

  function encodeTypedArray(et: ElementType, arr: ArrayBufferView): Node {
    const bytes = new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength);
    return leaf((w) => {
      w.u8(Tag.TypedArray);
      w.u8(et);
      w.uvarint(bytes.length);
      w.bytes(bytes);
    });
  }

  // Own enumerable string + symbol properties, shared by Object and Error.
  // `skip` excludes string keys handled out-of-band (Error's name/message/...).
  type Prop = { kind: KeyKind; key: string | number; val: number };
  function ownEntries(obj: object, skip?: Set<string>): Prop[] {
    const props: Prop[] = [];
    for (const k of Object.keys(obj)) {
      if (skip?.has(k)) continue;
      props.push({ kind: KeyKind.String, key: k, val: intern((obj as any)[k]) });
    }
    for (const s of Object.getOwnPropertySymbols(obj)) {
      if (!Object.getOwnPropertyDescriptor(obj, s)?.enumerable) continue;
      props.push({ kind: KeyKind.SymbolRef, key: intern(s), val: intern((obj as any)[s]) });
    }
    return props;
  }
  function writeEntries(w: ByteWriter, props: Prop[]): void {
    w.uvarint(props.length);
    for (const p of props) {
      w.u8(p.kind);
      if (p.kind === KeyKind.String) w.str(p.key as string);
      else w.uvarint(p.key as number);
      w.uvarint(p.val);
    }
  }
  // True when `key` is a canonical array index within [0, length).
  function isArrayIndex(key: string, length: number): boolean {
    const n = Number(key);
    return Number.isInteger(n) && n >= 0 && n < length && String(n) === key;
  }

  function buildObject(obj: object): Node {
    // Boxed primitives — unwrap to their primitive (best-effort; wrapper
    // identity and any extra props are dropped, see FORMAT.md §9).
    if (obj instanceof Number) return buildNumber(obj.valueOf());
    if (obj instanceof String) {
      const sv = obj.valueOf();
      return leaf((w) => {
        w.u8(Tag.String);
        w.str(sv);
      });
    }
    if (obj instanceof Boolean) {
      const bv = obj.valueOf();
      return leaf((w) => w.u8(bv ? Tag.BoolTrue : Tag.BoolFalse));
    }
    // Date — leaf, checked before structural types.
    if (obj instanceof Date) {
      const ms = obj.getTime();
      return leaf((w) => {
        w.u8(Tag.Date);
        w.svarint(BigInt(ms));
        w.svarint(0n); // sub_ms_nanos: JS は常に 0
      });
    }
    // TypedArrays — concrete subtype first, then ArrayBuffer, all before Array.
    if (obj instanceof BigInt64Array) return encodeTypedArray(ElementType.BigInt64, obj);
    if (obj instanceof BigUint64Array) return encodeTypedArray(ElementType.BigUint64, obj);
    if (obj instanceof Float64Array) return encodeTypedArray(ElementType.Float64, obj);
    if (obj instanceof Float32Array) return encodeTypedArray(ElementType.Float32, obj);
    if (obj instanceof Int32Array) return encodeTypedArray(ElementType.Int32, obj);
    if (obj instanceof Uint32Array) return encodeTypedArray(ElementType.Uint32, obj);
    if (obj instanceof Int16Array) return encodeTypedArray(ElementType.Int16, obj);
    if (obj instanceof Uint16Array) return encodeTypedArray(ElementType.Uint16, obj);
    if (obj instanceof Int8Array) return encodeTypedArray(ElementType.Int8, obj);
    if (obj instanceof Uint8ClampedArray) return encodeTypedArray(ElementType.Uint8Clamped, obj);
    if (obj instanceof Uint8Array) return encodeTypedArray(ElementType.Uint8, obj);
    if (obj instanceof ArrayBuffer) {
      const bytes = new Uint8Array(obj);
      return leaf((w) => {
        w.u8(Tag.Bytes);
        w.uvarint(bytes.length);
        w.bytes(bytes);
      });
    }
    // DataView — its own tag (NOT a TypedArray); stores the viewed window.
    if (obj instanceof DataView) {
      const bytes = new Uint8Array(obj.buffer, obj.byteOffset, obj.byteLength);
      return leaf((w) => {
        w.u8(Tag.DataView);
        w.uvarint(bytes.length);
        w.bytes(bytes);
      });
    }
    if (Array.isArray(obj)) {
      // Arrays carry only their elements 0..length-1. Extra own enumerable
      // properties cannot be represented, so reject rather than drop them.
      const extra = [
        ...Object.keys(obj).filter((k) => !isArrayIndex(k, obj.length)),
        ...Object.getOwnPropertySymbols(obj)
          .filter((s) => Object.getOwnPropertyDescriptor(obj, s)?.enumerable)
          .map((s) => String(s)),
      ];
      if (extra.length > 0) {
        throw new Error("array with non-index properties is not supported: " + extra.join(", "));
      }
      // Iterate by index so holes in sparse arrays become explicit undefined.
      const refs: number[] = new Array(obj.length);
      for (let i = 0; i < obj.length; i++) refs[i] = intern((obj as unknown[])[i]);
      return leaf((w) => {
        w.u8(Tag.Array);
        w.uvarint(refs.length);
        for (const r of refs) w.uvarint(r);
      });
    }
    if (obj instanceof Map) {
      const entries = [...obj.entries()].map(([k, val]) => [intern(k), intern(val)] as const);
      return leaf((w) => {
        w.u8(Tag.Map);
        w.uvarint(entries.length);
        for (const [k, val] of entries) {
          w.uvarint(k);
          w.uvarint(val);
        }
      });
    }
    if (obj instanceof Set) {
      const refs = [...obj.values()].map((x) => intern(x));
      return leaf((w) => {
        w.u8(Tag.Set);
        w.uvarint(refs.length);
        for (const r of refs) w.uvarint(r);
      });
    }
    if (obj instanceof WeakMap) {
      const supplied = provider.weakMapEntries?.(obj as WeakMap<object, unknown>) ?? [];
      const entries = supplied.map(([k, val]) => [intern(k), intern(val)] as const);
      return leaf((w) => {
        w.u8(Tag.WeakMap);
        w.uvarint(entries.length);
        for (const [k, val] of entries) {
          w.uvarint(k);
          w.uvarint(val);
        }
      });
    }
    if (obj instanceof WeakSet) {
      const supplied = provider.weakSetValues?.(obj as WeakSet<object>) ?? [];
      const refs = supplied.map((x) => intern(x));
      return leaf((w) => {
        w.u8(Tag.WeakSet);
        w.uvarint(refs.length);
        for (const r of refs) w.uvarint(r);
      });
    }
    // RegExp — leaf carrying its source pattern and flag string.
    if (obj instanceof RegExp) {
      return leaf((w) => {
        w.u8(Tag.RegExp);
        w.str(obj.source);
        w.str(obj.flags);
      });
    }
    // URL — leaf carrying the serialized href.
    if (typeof URL !== "undefined" && obj instanceof URL) {
      const href = obj.href;
      return leaf((w) => {
        w.u8(Tag.Url);
        w.str(href);
      });
    }
    // Error (and subclasses) — name + message + optional cause + extra props.
    // `stack` is environment-derived and intentionally not encoded.
    if (obj instanceof Error) {
      const name = obj.name;
      const message = obj.message;
      const hasCause = Object.prototype.hasOwnProperty.call(obj, "cause");
      const causeRef = hasCause ? intern((obj as { cause?: unknown }).cause) : 0;
      const extra = ownEntries(obj, new Set(["name", "message", "stack", "cause"]));
      return leaf((w) => {
        w.u8(Tag.Error);
        w.str(name);
        w.str(message);
        w.u8(hasCause ? 1 : 0);
        if (hasCause) w.uvarint(causeRef);
        writeEntries(w, extra);
      });
    }
    // User-registered extension types claim otherwise-unsupported objects.
    // The surrogate is interned as a child so it can be any Graft value.
    for (const t of types) {
      if (t.match(obj)) {
        const ref = intern(t.encode(obj));
        return leaf((w) => {
          w.u8(Tag.Custom);
          w.str(t.name);
          w.uvarint(ref);
        });
      }
    }
    // Guard: only plain records (Object.prototype / null prototype) may fall
    // through to the generic Object encoder. Any other exotic object would lose
    // its internal state when reduced to own enumerable props, so we reject it
    // loudly instead of silently emitting an empty/partial Object.
    const proto = Object.getPrototypeOf(obj);
    if (proto !== null && proto !== Object.prototype) {
      const name = obj.constructor?.name ?? "object";
      throw new Error("unsupported object type: " + name + " (no Graft tag for this exotic type)");
    }
    // Plain object: own enumerable string + symbol keys.
    const props = ownEntries(obj);
    return leaf((w) => {
      w.u8(Tag.Object);
      writeEntries(w, props);
    });
  }

  const rootId = intern(root);

  const w = new ByteWriter();
  w.bytes(MAGIC);
  w.u8(VERSION);
  w.uvarint(rootId);
  w.uvarint(heap.length);
  for (const node of heap) node.write(w, (v) => ids.get(v)!);
  return w.toUint8Array();
}

File: conformance/python/encode.py
"""Graft binary encoder — Python port.

The inverse of decode.py. It mirrors the reference JS encoder's algorithm
exactly — depth-first pre-order interning, identity dedup for objects and value
dedup for primitives — so that `encode(decode(bytes)) == bytes` for the golden
vectors (a byte-identical round-trip), even though byte equality is not required
for conformance in general (conformance/README.md §2).
"""

from __future__ import annotations

import struct
from typing import Any, Dict, List, Tuple

from decode import (
    BigInt,
    Bytes,
    DataView,
    Date,
    ELEMENT_TYPE_NAMES,
    GraftError,
    GraftMap,
    GraftSet,
    RegExp,
    Symbol,
    TypedArray,
    _Undefined,
    Url,
)

ELEMENT_TYPE_CODES = {name: code for code, name in ELEMENT_TYPE_NAMES.items()}
SYMBOL_TAG = {"registered": 10, "unique": 11, "well_known": 12}


class ByteWriter:
    def __init__(self) -> None:
        self.buf = bytearray()

    def u8(self, b: int) -> None:
        self.buf.append(b & 0xFF)

    def raw(self, data: bytes) -> None:
        self.buf.extend(data)

    def uvarint(self, n: int) -> None:
        if n < 0:
            raise ValueError("uvarint expects non-negative")
        while True:
            byte = n & 0x7F
            n >>= 7
            if n:
                self.buf.append(byte | 0x80)
            else:
                self.buf.append(byte)
                return

    def svarint(self, n: int) -> None:
        self.uvarint((n << 1) if n >= 0 else ((-n << 1) - 1))

    def f64(self, v: float) -> None:
        self.buf.extend(struct.pack("<d", v))

    def string(self, s: str) -> None:
        b = s.encode("utf-8")
        self.uvarint(len(b))
        self.buf.extend(b)

    def to_bytes(self) -> bytes:
        return bytes(self.buf)


# Reference types are deduped by object identity; primitives by value, matching
# the single identity/value map the JS encoder keys on.
_REF_TYPES = (
    list,
    dict,
    Symbol,
    GraftMap,
    GraftSet,
    GraftError,
    Date,
    Bytes,
    DataView,
    TypedArray,
    RegExp,
    Url,
)


def _is_ref(v: Any) -> bool:
    return isinstance(v, _REF_TYPES)


def _val_key(v: Any) -> Tuple:
    if v is None:
        return ("null",)
    if isinstance(v, _Undefined):
        return ("undef",)
    if isinstance(v, bool):
        return ("bool", v)
    if isinstance(v, BigInt):
        return ("bigint", str(int(v)))
    if isinstance(v, int):
        return ("int", v)
    if isinstance(v, float):
        return ("float", struct.pack("<d", v))
    if isinstance(v, str):
        return ("str", v)
    raise TypeError(f"not a primitive: {v!r}")


def encode(root: Any) -> bytes:
    heap: List[bytes] = []
    ids_obj: Dict[int, int] = {}  # id(obj) -> heap index
    ids_val: Dict[Tuple, int] = {}  # (kind, value) -> heap index

    def intern(v: Any) -> int:
        if _is_ref(v):
            key = id(v)
            if key in ids_obj:
                return ids_obj[key]
            idx = len(heap)
            ids_obj[key] = idx
            heap.append(b"")  # reserve index before building children (cycles)
            heap[idx] = build(v)
            return idx
        key2 = _val_key(v)
        if key2 in ids_val:
            return ids_val[key2]
        idx = len(heap)
        ids_val[key2] = idx
        heap.append(b"")
        heap[idx] = build(v)
        return idx

    def write_entries(w: ByteWriter, items) -> None:
        # JS order: string keys first (insertion order), then symbol keys; the
        # decoded dict already preserves that order. Refs are interned eagerly.
        parts: List[Tuple[int, Any, int]] = []
        for k, val in items:
            if isinstance(k, Symbol):
                parts.append((1, intern(k), intern(val)))
            else:
                parts.append((0, k, intern(val)))
        w.uvarint(len(parts))
        for kind, key, valref in parts:
            w.u8(kind)
            if kind == 0:
                w.string(key)
            else:
                w.uvarint(key)
            w.uvarint(valref)

    def build(v: Any) -> bytes:
        w = ByteWriter()
        if v is None:
            w.u8(0)
        elif isinstance(v, _Undefined):
            w.u8(1)
        elif isinstance(v, bool):
            w.u8(3 if v else 2)
        elif isinstance(v, BigInt):
            n = int(v)
            w.u8(6)
            w.u8(1 if n < 0 else 0)
            w.uvarint(abs(n))
        elif isinstance(v, int):
            w.u8(4)
            w.svarint(v)
        elif isinstance(v, float):
            w.u8(5)
            w.f64(v)
        elif isinstance(v, str):
            w.u8(7)
            w.string(v)
        elif isinstance(v, Symbol):
            w.u8(SYMBOL_TAG[v.kind])
            w.string(v.value)
        elif isinstance(v, Date):
            w.u8(40)
            w.svarint(v.unix_ms)
            w.svarint(v.sub_ms_nanos)
        elif isinstance(v, Bytes):
            w.u8(41)
            w.uvarint(len(v.data))
            w.raw(bytes(v.data))
        elif isinstance(v, TypedArray):
            w.u8(42)
            w.u8(ELEMENT_TYPE_CODES[v.element_type])
            w.uvarint(len(v.data))
            w.raw(bytes(v.data))
        elif isinstance(v, RegExp):
            w.u8(43)
            w.string(v.source)
            w.string(v.flags)
        elif isinstance(v, Url):
            w.u8(44)
            w.string(v.href)
        elif isinstance(v, DataView):
            w.u8(45)
            w.uvarint(len(v.data))
            w.raw(bytes(v.data))
        elif isinstance(v, list):  # Array
            refs = [intern(x) for x in v]
            w.u8(20)
            w.uvarint(len(refs))
            for r in refs:
                w.uvarint(r)
        elif isinstance(v, dict):  # Object
            w.u8(21)
            write_entries(w, v.items())
        elif isinstance(v, GraftMap):
            pairs = [(intern(k), intern(val)) for k, val in v.entries]
            w.u8(22)
            w.uvarint(len(pairs))
            for kr, vr in pairs:
                w.uvarint(kr)
                w.uvarint(vr)
        elif isinstance(v, GraftSet):
            refs = [intern(x) for x in v.values]
            w.u8(23)
            w.uvarint(len(refs))
            for r in refs:
                w.uvarint(r)
        elif isinstance(v, GraftError):
            cause_ref = intern(v.cause) if v.has_cause else 0
            w.u8(46)
            w.string(v.name)
            w.string(v.message)
            w.u8(1 if v.has_cause else 0)
            if v.has_cause:
                w.uvarint(cause_ref)
            write_entries(w, v.extra.items())
        else:
            raise TypeError(f"cannot encode {v!r}")
        return w.to_bytes()

    root_id = intern(root)
    out = ByteWriter()
    out.raw(b"GRF1")
    out.u8(1)
    out.uvarint(root_id)
    out.uvarint(len(heap))
    for node in heap:
        out.raw(node)
    return out.to_bytes()
